pkg/controller/endpoint/trigger_time_tracker.go

```diff
@@ -0,0 +1,219 @@
+		return
+	}
+
+	if triggerTime.Before(this.lastSyncMaxTriggerTime[key]) ||
+			triggerTime == this.lastSyncMaxTriggerTime[key] {
+		// Trigger was already processed
+		if triggerTime.Before(this.lastSyncMinTriggerTime[key]) {
+			// Oops, we exported a wrong time in the last processing. Increment the error counter.
+ 			LastChangeTriggerTimeMiscalculated.Inc()
+			klog.Warningf("Miscalculated LastChangeTriggerTime annotation for service %s. " +
+				"Should export: %s, exported %s", key, triggerTime, this.lastSyncMinTriggerTime[key])
+
+			// Correct the lastSyncMinTriggerTime so we don't increment the error again for the same
+			// batch. It could happen if the trigger change times were T0, T1, T2 and we exported T2 in
+			// the last sync. Later we would observe the T0 and T1 trigger change times and because both
+			// T0<T2 and T1 < T2 we would export the metric twice.
+			this.lastSyncMinTriggerTime[key] = triggerTime
+		}
+		return
+	}
+
+	// If we are here it means that the triggerTime is after lastSyncMaxTriggerTime. If this is first
+	// such triggerTime per batch, i.e. minTriggerTime is not set, then we should set it. If it was
+	// already set then we don't need to set it anymore for this batch because events are in the
+	// chronological order.
+	if _, ok := this.minTriggerTime[key]; !ok {
+		this.minTriggerTime[key] = triggerTime
+	}
+}
+
+// Method to be called directly before the call to the "list" function.
+func (this *triggerTimeTracker) startListing(key string) {
+	this.mutex.Lock()
+	defer this.mutex.Unlock()
+	this.isListingObjects[key] = true
+}
+
+// Method that should be called directly after the listing objects has ended. It will reset the
+// state for the given endpoints key and return the time that should be exported as the
+// EndpointsLastChangeTriggerTime annotation. It may happen that the method will return nil, in such
+// case the annotation shouldn't be exported.
+func (this *triggerTimeTracker) stopListingAndReset(
+		key string, triggerTimesFromListing []time.Time) (*time.Time){
+	this.mutex.Lock()
+	defer this.mutex.Unlock()
+
+	var endpointsLastChangeTriggerTime time.Time
+
+	if _, ok := this.minTriggerTime[key]; !ok {
+		// There was no event observed that set the minTriggerTime.
+
+		minTriggerTime := func() *time.Time {
+			// Try in the dirty times.
+			if p := minGreaterThan(this.dirtyTriggerTimes[key], this.lastSyncMaxTriggerTime[key]);
+					p != nil {
+				return p
+			}
+			// If nothing found in the dirty times, try in the times from listing objects.
+			if p := minGreaterThan(triggerTimesFromListing, this.lastSyncMaxTriggerTime[key]); p != nil {
+				return p
+			}
+
+			return nil
+		}()
+
+		if minTriggerTime == nil {
+			// Nothing was observed, nothing in the dirty set, nothing fresh enough in listing times.
+			// Reset state and return nil, there is nothing to export this time.
+			this.isListingObjects[key] = false
```

Comment:
Yes, we do also in the line 182, in other words we do that always when this method is called. 
Rewrote the method to make it more clear that this will be always called.